src/components/TextInput/TextInput.js

```diff
@@ -1,104 +1,190 @@
 import PropTypes from 'prop-types';
 import React from 'react';
 import classNames from 'classnames';
+import Icon from '../Icon';
 
-const TextInput = ({
-  labelText,
-  className,
-  id,
-  placeholder,
-  type,
-  onChange,
-  onClick,
-  hideLabel,
-  invalid,
-  invalidText,
-  helperText,
-  light,
-  ...other
-}) => {
-  const textInputProps = {
-    id,
-    onChange: evt => {
-      if (!other.disabled) {
-        onChange(evt);
-      }
-    },
-    onClick: evt => {
-      if (!other.disabled) {
-        onClick(evt);
-      }
-    },
-    placeholder,
-    type,
+export default class TextInput extends React.Component {
+  state = {
+    type: this.props.type,
   };
 
-  const errorId = id + '-error-msg';
-  const textInputClasses = classNames('bx--text-input', className, {
-    'bx--text-input--light': light,
-  });
-  const labelClasses = classNames('bx--label', {
-    'bx--visually-hidden': hideLabel,
-  });
-
-  const label = labelText ? (
-    <label htmlFor={id} className={labelClasses}>
-      {labelText}
-    </label>
-  ) : null;
-
-  const error = invalid ? (
-    <div className="bx--form-requirement" id={errorId}>
-      {invalidText}
-    </div>
-  ) : null;
+  togglePasswordVisibility = () => {
+    this.setState({
+      type: this.state.type === 'password' ? 'text' : 'password',
+    });
+  };
 
-  const input = invalid ? (
-    <input
-      {...other}
-      {...textInputProps}
-      data-invalid
-      aria-invalid
-      aria-describedby={errorId}
-      className={textInputClasses}
-    />
-  ) : (
-    <input {...other} {...textInputProps} className={textInputClasses} />
-  );
+  togglePasswordVisibilityIconProps = () => {
+    const passwordIsVisible = this.state.type === 'text';
+    return {
+      alt: `${passwordIsVisible ? 'Hide' : 'Show'} password`,
```

Comment:
A couple of things from g11n perspective here:

1. Should be customizable via `props` (in case you haven't done this)
2. Should avoid string interpolation. Avoiding string interpolation seems easy given we can prepare different strings for different states

Reply:
@asudoh I'm a bit confused about the string interpolation part. Would we expect users to provide strings for the two different button states?